### PR TITLE
Always use sha for sandbox image

### PR DESCRIPTION
**Pull request by evalsocket**
# TL;DR
- Currently we use dind tag for default sandbox image but we are planing to stop publishing dind image. In that case flytectl can query the latest tag and then create the SHA tag for sandbox image
- Added pre release logic
- Unit test hits the limit of github, Added auth for CI 

Tests
```
17:57:57 ➜ flytectl go run main.go sandbox start --version=v0.19.0
🧑‍🏭 Bootstrapping a brand new flyte cluster... 🔨 🔧
delete existing sandbox cluster [y/n]: y
🐋 Running Flyte v0.19.0 release
🐋 pulling docker image for release cr.flyte.org/flyteorg/flyte-sandbox:dind-1faf72e5f5ae59edb9647b2f3164271bb914e239
{"status":"Pulling from flyteorg/flyte-sandbox","id":"dind-1faf72e5f5ae59edb9647b2f3164271bb914e239"}
{"status":"Digest: sha256:5288c0bdabd80a0c1ab77526f23f2a855b60acff54e34a58d4c79474491490a2"}
{"status":"Status: Image is up to date for cr.flyte.org/flyteorg/flyte-sandbox:dind-1faf72e5f5ae59edb9647b2f3164271bb914e239"}
🧑‍🏭 booting Flyte-sandbox container
^Csignal: interrupt
```

```
17:58:46 ➜ flytectl go run main.go sandbox start                  
🧑‍🏭 Bootstrapping a brand new flyte cluster... 🔨 🔧
delete existing sandbox cluster [y/n]: y
🐋 Running Flyte v0.19.1 release
🐋 pulling docker image for release cr.flyte.org/flyteorg/flyte-sandbox:dind-362a3d33c56f65c3584ecdc0da48c2f149c9b075
{"status":"Pulling from flyteorg/flyte-sandbox","id":"dind-362a3d33c56f65c3584ecdc0da48c2f149c9b075"}
{"status":"Digest: sha256:9854de3d85b936d0ac7a299cf694f452bc3ad1b605d4408725d68da30b195492"}
{"status":"Status: Image is up to date for cr.flyte.org/flyteorg/flyte-sandbox:dind-362a3d33c56f65c3584ecdc0da48c2f149c9b075"}
🧑‍🏭 booting Flyte-sandbox container
^Csignal: interrupt
```

```
17:59:14 ➜ flytectl go run main.go sandbox start --pre            
🧑‍🏭 Bootstrapping a brand new flyte cluster... 🔨 🔧
delete existing sandbox cluster [y/n]: y
🐋 Running Flyte v0.19.2-b1 release
🐋 pulling docker image for release cr.flyte.org/flyteorg/flyte-sandbox:dind-1cac60d1d8233f48874257ca516dcc78c58b93ec
{"status":"Pulling from flyteorg/flyte-sandbox","id":"dind-1cac60d1d8233f48874257ca516dcc78c58b93ec"}
{"status":"Digest: sha256:3091a780a30ce08cdda29e716abb1dcbb6ead84ba51db5ed528acbb492611f65"}
{"status":"Status: Downloaded newer image for cr.flyte.org/flyteorg/flyte-sandbox:dind-1cac60d1d8233f48874257ca516dcc78c58b93ec"}
🧑‍🏭 booting Flyte-sandbox container
^Csignal: interrupt
```

## Type
- [ ] Bug Fix
- [x] Feature
- [ ] Plugin

## Are all requirements met?

- [x] Code completed
- [x] Smoke tested
- [x] Unit tests added
- [x] Code documentation added
- [ ] Any pending items have an associated Issue

## Complete description
_How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2080

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

**Comment by pmahindrakar-oss** [comment]
Please use the UX suggestions from this comment by Haytham
https://github.com/flyteorg/flyte/issues/2080#issuecomment-1018176807


Also it would be useful to know from the logs what were the commandline params being passed.
This might be generic ask for all but found that the current sandbox start logs dont tell v0.19.0 is being used when its passed in the commandline

Similarly if we use pre-release or latest would be good to know those aswell along with usual sha that gets printed. May be we can print the tags which have this info?

**Comment by evalsocket** [comment]
@pmahindrakar-oss added changes as discussed

File: cmd/sandbox/start.go [hunk 1]
```diff
@@ -188,31 +190,46 @@ func startSandbox(ctx context.Context, cli docker.Docker, reader io.Reader) (*bu
...
+		return "", err
+	}
+	if !isGreater {
+		return "", fmt.Errorf("version flag only supported with flyte %s+ release", sandboxSupportedVersion)
+	}
+	sha, err := githubutil.GetSHAFromVersion(*selectedVersion.TagName, flyteRepository)
+	if err != nil {
+		return "", err
+	}
+	return docker.GetSandboxImage(fmt.Sprintf("dind-%s", sha)), nil
```
Comment (pmahindrakar-oss) [inline cmd/sandbox/start.go:232]: L217-228 can all be abstracted out inside this function

**Comment by pmahindrakar-oss** [comment]
The test logs in the description show release being pulled even though you have passed in --preelease flag

File: cmd/sandbox/start.go [hunk 2]
```diff
@@ -188,31 +190,46 @@ func startSandbox(ctx context.Context, cli docker.Docker, reader io.Reader) (*bu
...
-		if !isGreater {
-			return "", fmt.Errorf("version flag only supported with flyte %s+ release", sandboxSupportedVersion)
-		}
-		sha, err := githubutil.GetSHAFromVersion(version, flyteRepository)
-		if err != nil {
-			return "", err
+		for _, v := range releases {
+			if *v.Prerelease && sandboxConfig.DefaultConfig.Prerelease {
+				logger.Infof(context.Background(), "sandbox started with pre release %s", *v.TagName)
+				version = *v.TagName
```
[the remark this answers is not in the document]
Comment (evalsocket) [reply cmd/sandbox/start.go:208]: yes

File: cmd/sandbox/start.go [hunk 3]
```diff
@@ -188,31 +190,46 @@ func startSandbox(ctx context.Context, cli docker.Docker, reader io.Reader) (*bu
...
 			return "", err
 		}
-		if !isGreater {
-			return "", fmt.Errorf("version flag only supported with flyte %s+ release", sandboxSupportedVersion)
-		}
-		sha, err := githubutil.GetSHAFromVersion(version, flyteRepository)
-		if err != nil {
-			return "", err
+		for _, v := range releases {
+			if *v.Prerelease && sandboxConfig.DefaultConfig.Prerelease {
```
[the remark this answers is not in the document]
Comment (evalsocket) [reply cmd/sandbox/start.go:206]: `sandboxConfig.DefaultConfig.Prerelease` is flag and other one is release status from list

File: cmd/sandbox/start.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (evalsocket) [reply cmd/sandbox/start.go:208]: yes also we are assuming that we will get it in first page of response

File: cmd/sandbox/start.go [hunk 4]
```diff
@@ -188,31 +190,46 @@ func startSandbox(ctx context.Context, cli docker.Docker, reader io.Reader) (*bu
...
+				logger.Infof(context.Background(), "sandbox started with pre release %s", *v.TagName)
+				version = *v.TagName
+				break
+			} else if !*v.Prerelease && !sandboxConfig.DefaultConfig.Prerelease {
```
[the remark this answers is not in the document]
Comment (evalsocket) [reply cmd/sandbox/start.go:210]: That is missing

File: pkg/util/githubutil/githubutil.go [hunk 5]
```diff
@@ -55,13 +55,23 @@ func GetGHClient() *github.Client {
...
 	if err != nil {
 		return nil, err
 	}
 	return release, err
 }
 
+// GetListRelease returns the list of release of provided repository
+func GetListRelease(repository string) ([]*github.RepositoryRelease, error) {
+	client := GetGHClient()
+	releases, _, err := client.Repositories.ListReleases(context.Background(), Owner, repository, &github.ListOptions{})
```
[the remark this answers is not in the document]
Comment (evalsocket) [reply pkg/util/githubutil/githubutil.go:68]: we will get latest first, sorted in datetime by default

File: cmd/sandbox/start.go [hunk 6]
```diff
@@ -188,31 +190,46 @@ func startSandbox(ctx context.Context, cli docker.Docker, reader io.Reader) (*bu
...
 		if err != nil {
 			return "", err
 		}
-		if !isGreater {
-			return "", fmt.Errorf("version flag only supported with flyte %s+ release", sandboxSupportedVersion)
-		}
-		sha, err := githubutil.GetSHAFromVersion(version, flyteRepository)
-		if err != nil {
-			return "", err
+		for _, v := range releases {
```
[the remark this answers is not in the document]
Comment (evalsocket) [reply cmd/sandbox/start.go]: done

**Comment by evalsocket** [comment]
@pmahindrakar-oss no it's correct, In case of pre release i am running pre release 
```
{"json":{"src":"start.go:206"},"level":"info","msg":"sandbox started with pre release v0.19.2-b1","ts":"2022-01-22T20:50:33+05:30"}

```

File: cmd/sandbox/start.go (hunk 1 above)
Comment (pmahindrakar-oss) [reply cmd/sandbox/start.go:201]: GetSandbox image can work directly off the version and you can even abstract the sha in there

File: cmd/sandbox/start.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply cmd/sandbox/start.go]: Please add this assumption as a note on the code block

**Comment by evalsocket** [comment]
@pmahindrakar-oss You can review the PR

**Comment by pmahindrakar-oss** [comment]
* For sha for sandbox image , I added one comment for hiding getSanboxImageSha function in the GetSandboxImage
```
	sha, err := githubutil.GetSandboxImageSha(version)
	if err != nil {
		return "", err
	}

	return docker.GetSandboxImage(tag), nil
	return docker.GetSandboxImage(fmt.Sprintf("dind-%s", sha)), nil
```
*Also while printing the log `sandbox started with release` to `sandbox using release %s` and similarly for pre-release
since sandbox hasn't started yet at this point and its just an image fetch


*Also this function looks out of place in githubutil
```
// IsVersionGreaterThan check version if it's greater then other
func IsVersionGreaterThan(version1, version2 string) (bool, error) {
	semanticVersion1, err := hversion.NewVersion(version1)
	if err != nil {
		return false, err
	}
	semanticVersion2, err := hversion.NewVersion(version2)
	if err != nil {
		return false, err
	}
	return semanticVersion1.GreaterThan(semanticVersion2), nil
}
```
This should be in versioningutil

* Also avoid using global config in deep functions
```
	if *v.Prerelease && sandboxConfig.DefaultConfig.Prerelease {
				logger.Infof(context.Background(), "sandbox started with pre release %s", *v.TagName)
				release = v
				break
			} else if !*v.Prerelease && !sandboxConfig.DefaultConfig.Prerelease {
				logger.Infof(context.Background(), "sandbox started with release %s", *v.TagName)
				release = v
				break
			}
```
Pass them as function params
* Also add a comment about the assumption being made that pre-release is part of the first 100 results from the ListApi for github.

File: cmd/config/subcommand/sandbox/sandbox_config.go [hunk 7]
```diff
@@ -44,6 +44,9 @@ type Config struct {
 	// from there.
 	Image string `json:"image" pflag:",Optional. Provide a fully qualified path to a Flyte compliant docker image."`
 
+	// Default value false represents that flytectl will not use latest pre release if exist
+	Prerelease bool `json:"pre" pflag:",Optional. Pre release Version of flyte will be used for sandbox."`
```
Comment (pmahindrakar-oss) [inline cmd/config/subcommand/sandbox/sandbox_config.go:48]: nit : can we also mention here which are other flags if passed in and there behavior eg : version or image

File: cmd/sandbox/start.go [hunk 8]
```diff
@@ -160,20 +158,23 @@ func startSandbox(ctx context.Context, cli docker.Docker, reader io.Reader) (*bu
 	} else if vol != nil {
 		volumes = append(volumes, *vol)
 	}
-
-	image, err := getSandboxImage(sandboxConfig.DefaultConfig.Version, sandboxConfig.DefaultConfig.Image)
-	if err != nil {
-		return nil, err
+	sandboxImage := sandboxConfig.DefaultConfig.Image
+	if len(sandboxConfig.DefaultConfig.Image) == 0 {
```
Comment (pmahindrakar-oss) [inline cmd/sandbox/start.go:162]: len(sandboxImage) == 0

File: pkg/util/githubutil/githubutil.go [hunk 9]
```diff
@@ -159,3 +211,15 @@ func CheckBrewInstall(goos platformutil.Platform) (string, error) {
 	}
 	return "", nil
 }
+
+// GetSandboxImage Returns the alternate image if specified, else
```
Comment (pmahindrakar-oss) [inline pkg/util/githubutil/githubutil.go:215]: This statement is not relevant to this function.

File: pkg/util/githubutil/githubutil.go [hunk 10]
```diff
@@ -159,3 +211,15 @@ func CheckBrewInstall(goos platformutil.Platform) (string, error) {
...
+// if no version is specified then the Latest release of cr.flyte.org/flyteorg/flyte-sandbox:dind-{SHA} is used
+// else cr.flyte.org/flyteorg/flyte-sandbox:dind-{SHA}, where sha is derived from the version.
+func GetSandboxImage(version, image string) (string, string, error) {
```
Comment (pmahindrakar-oss) [inline pkg/util/githubutil/githubutil.go:218]: not : GetSandBoxImage -> GetFullyQualifiedImageName
pass in the bool flag of pre-release

**Review by pmahindrakar-oss — APPROVED** [review]
